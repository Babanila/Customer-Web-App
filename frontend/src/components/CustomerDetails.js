import React from 'react'
import { useHistory } from 'react-router-dom'
import { cx, css } from 'emotion'
import FormComponent from './FormComponent'
import ErrorComponent from './ErrorComponent'
import SingleButton from './SingleButton'
import BackButton from './BackButton'
import { isoDateConverter } from './Utils'

const customerDetailsDivStyle = css`
  margin-top: 3em;
  width: 400px;
  height: auto;
  display: flex;
  align-self: center;
  flex-direction: column;
  border: 1px solid #f54b3a;
  border-radius: 5px;
`

const customerDetailsDivTitle = css`
  align-self: center;
  font-size: 2em;
  padding: 20px 0px;
`

const customerDetailsBtnGroup = css`
  margin-bottom: 20px;
  width: 42%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
  outline: none;
  background-color: #ffffff;
  color: #000000;
  font-size: 1.2em;
`

const updateBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  color: #009900;
  border: 2px solid #009900;
  background-color: #ffffff;
  border-radius: 5px;
  outline: none;
  &:hover {
    cursor: pointer;
    color: #ffffff;
    background-color: #009900;
  }
`

function CustomerDetails({ allCustomerData, match }) {
  const history = useHistory()
  const selectedCustomer = allCustomerData.find(
    (item) => item.customerID === parseInt(match.params.id)
  )

  const [changeCustomerData, setChangeCustomerData] = React.useState(selectedCustomer)
  const handleDataChange = ({ target }) => {
    const { name, value } = target
    let modifiedValue
    name === 'lastContact' ? (modifiedValue = isoDateConverter(value)) : (modifiedValue = value)
    setChangeCustomerData((state) => ({
      ...state,
      name: {
        ...state.name,
        first: name === 'first' ? modifiedValue : state.name.first,
        last: name === 'last' ? modifiedValue : state.name.last
      },
      [name]: modifiedValue
    }))
  }

  const handleUpdate = (e) => {
    e.preventDefault()
    console.log('Update button clicked')
  }

  const handleBack = () => history.push(`/`)

  const selectedCustomerDetails = (
    <div className={cx(customerDetailsDivStyle)}>
      <div className={cx(customerDetailsDivTitle)}>Customer Full Details</div>
      <FormComponent
        formData={changeCustomerData}
        handleSubmit={handleUpdate}
        handleInputChange={handleDataChange}
      />
      <div className={cx(customerDetailsBtnGroup)}>
        <SingleButton btnName="Update" btnClick={handleUpdate} btnStyles={updateBtnStyle} />
        <BackButton handleBack={handleBack} />
      </div>
    </div>
  )

  return selectedCustomer ? selectedCustomerDetails : <ErrorComponent />
}

export default CustomerDetails
