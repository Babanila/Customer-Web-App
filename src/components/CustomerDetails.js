import React from 'react';
import moment from 'moment';
import { cx, css } from 'emotion';

const tableRowDiv = css`
  display: table-row;
  background-color: #f6f6f6;
  &:hover {
    background-color: #ffe7e7;
  }
`;

const tableCellDiv = css`
  display: table-cell;
  text-align: center;
  padding: 10px;
  border-bottom: 1px solid black;
`;

function CustomerDetails({ customerData, onClick }) {
  const { customerID, name, lastContact, customerLifetimeValue } = customerData;
  return (
    <div className={cx(tableRowDiv)} onClick={() => onClick(customerID)}>
      <div className={cx(tableCellDiv)}>{customerID}</div>
      <div className={cx(tableCellDiv)}>{name.first}</div>
      <div className={cx(tableCellDiv)}>{name.last}</div>
      <div className={cx(tableCellDiv)}>{moment(lastContact).format('lll')}</div>
      <div className={cx(tableCellDiv)}>{customerLifetimeValue}</div>
    </div>
  );
}

export default CustomerDetails;
