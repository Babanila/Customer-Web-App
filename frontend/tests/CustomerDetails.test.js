import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import CustomerDetails from '../src/components/CustomerDetails'

describe('Check CustomerDetails Component', () => {
  let sampleData, CustomerDetailsWrapper, match
  let actualDiv, actualDiv1, actualDiv2

  beforeEach(() => {
    sampleData = {
      customerID: 4,
      name: {
        first: 'Roxy',
        last: 'Fox'
      },
      birthday: '1979-06-30',
      gender: 'w',
      lastContact: '2017-01-29T21:08:50.700Z',
      customerLifetimeValue: 213.12
    }

    match = { params: { id: 4 } }

    CustomerDetailsWrapper = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerDetails match={match} />
      </MemoryRouter>
    )

    actualDiv = CustomerDetailsWrapper.find('MemoryRouter')
    actualDiv1 = actualDiv.find('Router')
    actualDiv2 = actualDiv1.find('CustomerDetails')
  })

  it('should renders component without crashing', () => {
    const tree = renderer.create(<CustomerDetails match={match} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render CustomerDetails component without data', () => {
    const actualDiv3 = actualDiv2.find('LoadingComponent')

    expect(CustomerDetails).toBeDefined()
    expect(actualDiv.length).toEqual(1)
    expect(actualDiv1.length).toEqual(1)
    expect(actualDiv2.length).toEqual(1)
    expect(actualDiv3.length).toEqual(1)
  })

  it('should render CustomerDetails component data', () => {
    const setSelectedCustomer = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setSelectedCustomer(sampleData))

    const setLoading = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setLoading(false))

    const CustomerDetailsWrap = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerDetails match={match} />
      </MemoryRouter>
    )

    const actual1 = CustomerDetailsWrap.find('div')
    const actual2 = CustomerDetailsWrap.find('FormComponent')
    const actual3 = CustomerDetailsWrap.find('SingleButton')
    const actual4 = CustomerDetailsWrap.find('BackButton')
    const actual5 = actual2.children()

    expect(actual1.length).toEqual(13)
    expect(actual2.length).toEqual(1)
    expect(actual3.length).toEqual(3)
    expect(actual4.length).toEqual(1)
    expect(actual5.at(0).childAt(0).text()).toEqual('CustomerId:')
  })
})
