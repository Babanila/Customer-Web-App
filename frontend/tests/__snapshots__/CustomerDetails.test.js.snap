// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CustomerDetails Component should renders component without crashing 1`] = `
<h3
  className="css-d8gwku"
>
  Loading ...
</h3>
`;
